static/highlight.js

```diff
@@ -0,0 +1,83 @@
+// Copyright 2018 The Rustw Project Developers.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+import React from 'react';
+
+export const Highlight = (props) => {
+    const src_line_prefix = 'src_line_';
+    const highlight = props.highlight;
+    const is_valid_highlight = validate_highlight(highlight);
+
+    if (!is_valid_highlight) {
+        return null;
+    }
+
+    const lhs = (highlight.column_start - 1);
+    const rhs = (highlight.column_end - 1);
+    const highlight_specs = make_highlight(src_line_prefix, highlight.line_start, lhs, rhs);
+    if (highlight_specs) {
+        const { top, left, width } = highlight_specs;
+        const style = {
+            top: top,
+            left: left,
+            width: width,
+        }
+        return <div className="selected floating_highlight" key={highlight.line_start} style={style}>&nbsp;</div>;
+    }
+
+    return null;
+}
+
+function make_highlight(src_line_prefix, line_number, left, right) {
+    const line_div = $("#" + src_line_prefix + line_number);
+
+    // TODO: get adjust variable as prop through diffIndent in FileResult
+    // if Highlight component is to be used in the SearchResults component
+    // const adjust = line_div.data('adjust');
+    // if (adjust) {
+    //     left -= adjust;
+    //     right -= adjust;
+    // }
+
+    left *= CHAR_WIDTH;
+    right *= CHAR_WIDTH;
+    if (right === 0) {
+        right = line_div.width();
+    }
+
+    let width = right - left;
+    const paddingLeft = parseInt(line_div.css("padding-left"));
+    const paddingTop = parseInt(line_div.css("padding-top"));
+    if (left >= 0) {
+        left -= paddingLeft;
+    } else {
```

Comment:
We call it here: https://github.com/nrc/cargo-src/pull/237/files/ac2f86ab6bc765177f3e12af715b890c5de80476#diff-4018adfb8de39d7a30dd31dd44e3b4cdR22 and lhs is highlight.column - 1, so I think that should always be greater than 0 since the column should be 1-indexed, but we might want to check that it really is 1-indexed, not 0-indexed, and that seems like an easy mistake to make so it might be worth asserting that highlight.column >= 1 there.

Reply:
I found a case when testing a full-line highlight where `lhs` ends up being 0, and without changing the `left > 0` (as it is in [`utils.js`](https://github.com/nrc/cargo-src/blob/master/static/utils.js#L161)) to `left >= 0`, the highlight calculation is off slightly. But it sounds like we could just adjust the `validate_highlight` function to ensure `highlight.column_start` is not <= 1 rather than 0, as [it is currently](https://github.com/nrc/cargo-src/pull/237/files/ac2f86ab6bc765177f3e12af715b890c5de80476#diff-4018adfb8de39d7a30dd31dd44e3b4cdR79)?

Reply:
@nrc Ok, I'm re-thinking my previous comment - I think the `validate_highlight` logic can be left as-is, but the conditional adjustment in `make_highlight` could be reduced to `left -= paddingLeft`, with no check for `left >= 0` or any `else` clause. At least that's working for the test cases I've been running through.